elastica/boundary_conditions.py

```diff
@@ -406,9 +410,7 @@ def nb_constrain_translational_rates(velocity_collection, indices) -> None:
         block_size = indices.size
         for i in range(block_size):
             k = indices[i]
-            velocity_collection[0, k] = 0.0
-            velocity_collection[1, k] = 0.0
-            velocity_collection[2, k] = 0.0
+            velocity_collection[..., k] = 0.0
```

Comment:
Can we revert this change back? It includes extra instructor steps to broadcast and is typically faster to represent by unrolling vector. The above `position_collection[...,k] = fixed_position_collection[...,i]` is still okay because the number of instructions in the copying vector is probably the same.

It is a bit controversial topic since this step probably affects minimally on overall walltime, and it probably doesn't affect much. However, most of the old kernels we have are often tested multiple times with multiple implementations, and we don't want to change them for code simplification. Timing is a bit sensitive topic in our project, so please understand.

## code

``` py
import numpy as np
from numba import njit

N = 100
vec = np.random.random((3, N))

# --------------- Test Set Zero -------------- #
def f1(a):
    block_size = a.shape[1]
    for i in range(block_size):
        a[0,i] = 0.0
        a[1,i] = 0.0
        a[2,i] = 0.0
    
def f2(a):
    block_size = a.shape[1]
    for i in range(block_size):
        a[...,i] = 0.0
```

## result (N=100)

```py
In [1]: run test.py

In [2]: %timeit f1(vec)
27.3 µs ± 1.08 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [3]: %timeit f2(vec)
37.6 µs ± 722 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

## result (N=10000)

```py
In [1]: run test.py

In [2]: %timeit f1(vec)
2.8 ms ± 59.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [3]: %timeit f2(vec)
4.17 ms ± 121 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Reply:
@skim0119 can you post the timings with `njit` operator applied?

Reply:
```py
In [1]: run test.py

In [2]: f1(vec); f2(vec); # Cache, First run

In [3]: %timeit -r 10 f1(vec)
261 ns ± 1.24 ns per loop (mean ± std. dev. of 10 runs, 1000000 loops each)

In [4]: %timeit -r 10 f2(vec)
515 ns ± 9.93 ns per loop (mean ± std. dev. of 10 runs, 1000000 loops each)
```

Reply:
Ok, I feel 0.5 us is hardly gonna affect the rod timestep timing which for 100 elements is roughly 50-60 us, but I agree with @skim0119 regarding reverting the change since it does not lead to that much of a code simplification.

Reply:
@bhosale2 `numba` helps compile the code, but it probably doesn't optimize instruction itself, further than what LLVM does. For LLVM, we manually need to add `#pragma unroll` to specify SIMD vectorization, so I guess similar logic applies here.

Reply:
Thanks for the explanation. That is indeed quite interesting. I reverted the changes in https://github.com/GazzolaLab/PyElastica/pull/147/commits/9a95c4389de20fce48cc9771faeb714f4694607c